crates/turborepo-lib/src/run/watch.rs

```diff
@@ -1,42 +1,217 @@
             can_kill_server: true,
-            paths: DaemonPaths::from_repo_root(repo_root),
+            paths: DaemonPaths::from_repo_root(&base.repo_root),
         };
 
         let mut client = connector.connect().await?;
 
-        let mut hashes = client.package_changes().await?;
-        while let Some(hash) = hashes.next().await {
-            // Should we recover here?
-            let hash = hash.unwrap();
-            let event = hash.event.expect("event is missing");
-            match event {
-                proto::package_change_event::Event::PackageChanged(proto::PackageChanged {
-                    package_name,
-                }) => {
-                    println!("{} changed", package_name);
+        let mut events = client.package_changes().await?;
+        let mut current_runs: HashMap<PackageName, JoinHandle<Result<i32, run::Error>>> =
+            HashMap::new();
+        let event_fut = async {
+            while let Some(event) = events.next().await {
+                let event = event.unwrap();
+                Self::handle_change_event(
+                    &mut run,
+                    event.event.unwrap(),
+                    &mut current_runs,
+                    &base,
+                    &telemetry,
+                    &handler,
+                )
+                .await?;
+            }
+
+            Err(Error::ConnectionClosed)
+        };
+
+        select! {
+            biased;
+            _ = signal_subscriber.listen() => {
+                tracing::info!("shutting down");
+
+                Err(Error::SignalInterrupt)
+            }
+            result = event_fut => {
+                result
+            }
+        }
+    }
+
+    async fn handle_change_event(
+        run: &mut Run,
+        event: proto::package_change_event::Event,
+        current_runs: &mut HashMap<PackageName, JoinHandle<Result<i32, run::Error>>>,
+        base: &CommandBase,
+        telemetry: &CommandEventBuilder,
+        handler: &SignalHandler,
+    ) -> Result<(), Error> {
+        // Should we recover here?
+        match event {
+            proto::package_change_event::Event::PackageChanged(proto::PackageChanged {
+                package_name,
+            }) => {
+                let package_name = PackageName::from(package_name);
+                // If not in the filtered pkgs, ignore
+                if !run.filtered_pkgs.contains(&package_name) {
+                    return Ok(());
                 }
-                proto::package_change_event::Event::RediscoverPackages(_) => {
-                    println!("Rediscovering packages");
+
+                let mut args = base.args().clone();
+                args.command = args.command.map(|c| {
+                    if let Command::Watch(execution_args) = c {
+                        Command::Run {
+                            execution_args: Box::new(ExecutionArgs {
+                                filter: vec![format!("...{}", package_name)],
```

Comment:
I don't think this is the right filter? This will pull in everything that depends on the changed package, whether it was in the original run or not.